### PR TITLE
convert react-headroom to typescript to unblock migrations using ts-node

### PR DESCRIPTION
**Pull request by b0b3rt**
tl;dr: adding `import { EA_FORUM_HEADER_HEIGHT } from './common/Header';` to `Layout.tsx` borked migrations because `migrate.js` uses `ts-node` to import the server code, which has some differences in behavior from `esbuild`, and a bunch of transitive dependencies caused that new import to bring in `react-headroom`, which was a vendored JSX (rather than TSX) file.

This PR only converts `react-headroom` to Typescript, with no functional changes (i.e. untangling those imports, which can be done in a follow-up PR).

For more detail, see this slack thread: https://app.slack.com/client/T3ERD8FQT/CJUN2UAFN

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1204974955468522) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/react-headroom/index.tsx [hunk 1]
```diff
@@ -124,10 +148,11 @@ export default class Headroom extends Component {
...
   getScrollY = () => {
-    if (this.props.parent().pageYOffset !== undefined) {
-      return this.props.parent().pageYOffset
-    } else if (this.props.parent().scrollTop !== undefined) {
-      return this.props.parent().scrollTop
+    const parent = this.props.parent()
+    if ('pageYOffset' in parent && parent.pageYOffset !== undefined) {
+      return parent.pageYOffset
+    } else if ('scrollTop' in parent && parent.scrollTop !== undefined) {
+      return parent.scrollTop
```
Comment (b0b3rt) [inline packages/lesswrong/lib/react-headroom/index.tsx:155]: This would've been more annoying to type with `AnyBecauseTodo` than just making the language perform the correct type narrowing itself.  It should be functionally equivalent to the previous code (there'll never be a case where  `parent.pageYOffset !== undefined` returns `true` but `pageYOffset in parent` returns false).

File: packages/lesswrong/lib/react-headroom/index.tsx [hunk 2]
```diff
@@ -166,18 +191,18 @@ export default class Headroom extends Component {
 
     return (parent === window || parent === document.body)
       ? this.getViewportHeight()
-      : this.getElementPhysicalHeight(parent)
+      : this.getElementPhysicalHeight(parent as HTMLElement)
```
Comment (b0b3rt) [inline packages/lesswrong/lib/react-headroom/index.tsx:194]: This, however, was easier to forcibly type cast than leverage language-native type-narrowing functionality.

File: packages/lesswrong/lib/react-headroom/index.tsx [hunk 3]
```diff
@@ -289,14 +316,14 @@ export default class Headroom extends Component {
...
-    let innerStyle = {
+    let innerStyle: CSSProperties = {
       position: this.props.disable || this.state.state === 'unfixed' ? 'relative' : 'fixed',
       top: 0,
       left: 0,
       right: 0,
       zIndex: 1,
       WebkitTransform: `translate3D(0, ${this.state.translateY}, 0)`,
-      MsTransform: `translate3D(0, ${this.state.translateY}, 0)`,
+      msTransform: `translate3D(0, ${this.state.translateY}, 0)`,
```
Comment (b0b3rt) [inline packages/lesswrong/lib/react-headroom/index.tsx:326]: `csstype` complained about the capitalization here.  See this [thread](https://github.com/facebook/react/issues/2950) for more context.

**Review by s-cheng — APPROVED** [review]
Thank you! ❤️
